src/Microsoft.SqlTools.ServiceLayer/Agent/Jobs/AgentNotebookHelper.cs

```diff
@@ -567,79 +556,143 @@ UPDATE notebooks.nb_template
+        /// added for failed jobs which do not have an entry into the materialized table
+        /// </summary>
+        /// <param name="connInfo">connectionInfo generated from OwnerUri</param>
+        /// <param name="agentNotebookHistory">actual history item to be pinned</param>
+        /// <param name="targetDatabase">database on which the notebook history is stored</param>
+        /// <param name="pin">pin state for the history</param>
+        /// <returns></returns>
         public static async Task UpdateMaterializedNotebookPin(
             ConnectionInfo connInfo,
-            string materializedId,
+            AgentNotebookHistoryInfo agentNotebookHistory,
             string targetDatabase,
             bool pin)
         {
             string updateMaterializedNotebookPinQuery =
-             @"
+            @"
+            IF EXISTS
+            (SELECT * FROM notebooks.nb_materialized 
+            WHERE job_id = @jobId AND run_time = @startTime AND run_date = @startDate)
+            BEGIN
                 UPDATE notebooks.nb_materialized 
                 SET 
-                pin = @notebookPin 
+                pin = @notebookPin
                 WHERE 
-                materialized_id = @materializedId
+                job_id = @jobId AND run_time = @startTime AND run_date = @startDate
+            END
+            ELSE
+            BEGIN
+                INSERT INTO notebooks.nb_materialized (job_id, run_time, run_date, notebook, notebook_error, pin) 
+                VALUES 
+                (@jobID, @startTime, @startDate, '', '', @notebookPin)
+            END
             ";
             List<SqlParameter> updateMaterializedNotebookPinParams = new List<SqlParameter>();
+            updateMaterializedNotebookPinParams.Add(new SqlParameter("jobID", agentNotebookHistory.JobId));
+            updateMaterializedNotebookPinParams.Add(new SqlParameter("startTime", agentNotebookHistory.RunDate.ToString("HHmmss")));
+            updateMaterializedNotebookPinParams.Add(new SqlParameter("startDate", agentNotebookHistory.RunDate.ToString("yyyyMMdd")));
             updateMaterializedNotebookPinParams.Add(new SqlParameter("notebookPin", pin));
-            updateMaterializedNotebookPinParams.Add(new SqlParameter("materializedId", materializedId));
             await AgentNotebookHelper.ExecuteSqlQueries(
                 connInfo,
                 updateMaterializedNotebookPinQuery,
                 updateMaterializedNotebookPinParams,
                 targetDatabase);
         }
+
+        /// <summary>
+        /// Delete a particular run of the job. Deletion mainly including clearing out the notebook,
+        /// and notebook-error. The API doesn't delete the row because some notebook runs that have job
+        /// error in them don't have an entry in the materialized table. For keeping track of those notebook
+        /// runs the entry is added into the table with is_delete set to 1.
```

Comment:
Why do we want to keep track of these? If we're deleting them isn't that implying we don't care about them anymore?

Reply:
The API only deletes the artifacts created by notebook job history and not the history of the notebook job. So if the job has an error there won't be any entry in the materialized table. Therefore, just to remove the job from the UI I'm setting the 'is_deleted to false.